### DISCUSSION TITLE
How to get all variants and it's options?

### DISCUSSION THREAD

**Discussion by syabro** [joe-bell/cva#146] [Q&A] [answered]
E.g. I have

```ts
export const buttonVariants = cva(['your', 'base', 'classes'], {
  variants: {
    color: {
      primary: 'bg-primary-600 hover:bg-primary-500 text-white',
      dark: 'bg-gray-800 hover:bg-gray-700 text-white',
      light: 'bg-gray-200 hover:bg-gray-300 text-gray-800',
      danger: 'bg-red-600 hover:bg-red-500 text-white',
      gray: 'bg-gray-600 hover:bg-gray-500 text-white',
      info: 'bg-blue-600 hover:bg-blue-500 text-white',
      purple: 'bg-purple-600 hover:bg-purple-500 text-white',
      success: 'bg-green-600 hover:bg-green-500 text-white',
      warning: 'bg-yellow-600 hover:bg-yellow-500 text-white',
    },
    size: {
      xs: 'px-2 py-1',
      sm: 'px-2 py-1',
      md: 'px-2.5 py-1.5',
      lg: 'px-3 py-2',
      xl: 'px-4 py-2',
    },
  },
  defaultVariants: {
    color: 'primary',
    size: 'md',
  },
})
```

And I want to create a storybook with all options without copypasting them.
How can I extract all colors and sizes?

**Comment by syabro** [comment]
Ended up writing my own `tcva` function
Maybe we can add this functionality to the main package?

```ts
import { ClassValue, ClassProp, StringToBoolean } from 'class-variance-authority/dist/types'
import { ConfigSchema, ConfigVariants, Config, cva } from 'class-variance-authority'

type TcvaProps<T> = T extends ConfigSchema ? ConfigVariants<T> & ClassProp : ClassProp
type TcvaReturnFunction<T extends ConfigSchema> = {
  (props?: TcvaProps<T> | undefined): string
} & {
  [K in keyof T]: T[K]
}
export declare type TCVA = <T extends ConfigSchema>(
  base?: ClassValue,
  config?: Config<T> | undefined
) => TcvaReturnFunction<T>

export const tcva = <T extends ConfigSchema>(base: ClassValue, config: Config<T>) => {
  const variantsFn: TcvaReturnFunction<T> = cva(base, config) as TcvaReturnFunction<T>
  for (const variant of Object.keys(config.variants!)) {
    ;(variantsFn as any)[variant] = config.variants![variant]
  }
  return variantsFn
}

```

**Reply by switz** [reply]
do you have an updated version for the latest cva? seems some of these types are no longer exported

**Comment by joe-bell** [comment]
It's on my todo list for 1.0!

Thanks for sharing your solution 🙏

**Reply by syabro** [reply]
@joe-bell should I add a PR?

**Reply by joe-bell** [reply]
I appreciate the offer but that won't be necessary – there's some wider strategy to solve first

Thanks again!

**Reply by bengry** [reply]
All of the below is using `cva@1.0.0-beta.1` 👇

---

@joe-bell is this still on-track for 1.0? I think ideally we'd get something like:
```tsx
export const button = cva({
  base: ["your", "base", "classes"],
  variants: {
    color: {
      primary: "bg-primary-600 hover:bg-primary-500 text-white",
      dark: "bg-gray-800 hover:bg-gray-700 text-white",
      light: "bg-gray-200 hover:bg-gray-300 text-gray-800",
      danger: "bg-red-600 hover:bg-red-500 text-white",
      gray: "bg-gray-600 hover:bg-gray-500 text-white",
      info: "bg-blue-600 hover:bg-blue-500 text-white",
      purple: "bg-purple-600 hover:bg-purple-500 text-white",
      success: "bg-green-600 hover:bg-green-500 text-white",
      warning: "bg-yellow-600 hover:bg-yellow-500 text-white",
    },
    size: {
      xs: "px-2 py-1",
      sm: "px-2 py-1",
      md: "px-2.5 py-1.5",
      lg: "px-3 py-2",
      xl: "px-4 py-2",
    },
  },
  defaultVariants: {
    color: "primary",
    size: "md",
  },
});
```
```tsx
// usage
button.variants // => ['color', 'size']
button({ ... }) // "your base classes..."
```

The implementation is quite trivial (in both JS and types), but I was not able to wrap `cva` itself while retaining the types, since the way it's typed is very internal, for lack of a better term. So the only course of action is to [patch](https://www.npmjs.com/package/patch-package) `cva` itself, which I'd rather like to avoid if possible.

**Reply by joe-bell** [reply]
Would you be open to sharing your patch @bengry? I’d love to see it!

**Reply by bengry** [reply]
@joe-bell Here's a rather quick implementation of it.
~https://github.com/bengry/cva/commit/aed3d89?w=1~
https://github.com/joe-bell/cva/compare/main...bengry:cva:variants-return-property?w=1

LMK if you have any questions.

~Note that I did apparently break some of `compose`'s types. This is why I had to commit by skipping the pre-commit hooks. I'm certain we can get those working as well pretty easily, at least at a basic level where `compose` doesn't break, even if it doesn't support the `variants` property itself. I mainly wanted to hack it together quickly for now.~
I added some minor test cases to make sure nothing breaks and to POC this approach, those can surely be expanded. Also, naming could probably be better - but I just wanted to hack something together to see if this is something that you'd be interested in building upon/merging.

The main use-case, by the way, at least for me is to stop having to destructure props in my components. i.e. I have something like:
```tsx
import { cva, VariantProps } from "../utils/cva";

export const Heading: React.FC<
  VariantProps<typeof heading> & {
    className?: string;
    as?: "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
  } & React.HTMLAttributes<HTMLHeadingElement>
> = ({ color, weight, className, children, as: Tag = "h2", ...props }) => {
  return (
    <Tag className={heading({ color, weight, className })} {...props}>
      {children}
    </Tag>
  );
};

const heading = cva({
  base: ["text-current-hue-12", "tracking-tight"],
  variants: {
    weight: {
      light: "font-light",
      normal: "font-normal",
      bold: "font-bold",
    },
    color: {
      default: "hue-gray",
      red: "hue-red",
      green: "hue-green",
      accent: "hue-blue",
      yellow: "hue-yellow",
    },
  },
});
```

If I add another variant now, I have to remember to go back to the component and destructure it as well and pass it into `heading`. Instead, I want to do this (or similar):
```tsx
import { cva, VariantProps } from "../utils/cva";
import { pick, omit } from "lodash";

export const Heading: React.FC<
  VariantProps<typeof heading> & {
    className?: string;
    as?: "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
  } & React.HTMLAttributes<HTMLHeadingElement>
> = ({ className, children, as: Tag = "h2", ...props }) => {
  return (
    <Tag
      {...omit(props, heading.variants)}
      className={heading({
        ...pick(props, heading.variants),
        className,
      })}
    >
      {children}
    </Tag>
  );
};

const heading = cva({
  base: ["text-current-hue-12", "tracking-tight"],
  variants: {
    weight: {
      light: "font-light",
      normal: "font-normal",
      bold: "font-bold",
    },
    color: {
      default: "hue-gray",
      red: "hue-red",
      green: "hue-green",
      accent: "hue-blue",
      yellow: "hue-yellow",
    },
  },
});
```

**Reply by TomFreudenberg** [reply]
Hi @bengry @joe-bell 

is there an update about implementation / decision for `.variants` property.

I am highly interested in and wonder about release.

Currently using {"class-variance-authority": "^0.7.0"}

Thanks for some feedback
Tom

**Reply by nikuscs** [reply]
Same, specially in vue that we cannot use complex types to infer as component types :(

**Reply by bengry** [reply]
@TomFreudenberg From my point of view, I implemented this in https://github.com/joe-bell/cva/compare/main...bengry:cva:variants-return-property.
I never got an ack from @joe-bell that they're interested in adding this to cva, so I haven't created a PR for this. But I'd be happy to polish things up and create a PR if there's an intention to merge this from @joe-bell.

**Reply by joe-bell** [reply]
I’m interested, I’m just personally at capacity right now — introducing new features like this requires a lot of time to think and strategise

As much as I love this project and want to push it forwards, is not my number one priority 🙏🏼

**Reply by bengry** [reply]
@joe-bell Thanks for the prompt response.

>I’m interested, I’m just personally at capacity right now — introducing new features like this requires a lot of time to think and strategise

Could you please expand on this some more?

Also, did you have a chance to look at my proposed code changes ([this](https://github.com/joe-bell/cva/compare/0bc637014dbe5785481f351c61f1c311d561ed05...bengry:cva:variants-return-property?w=0) is a more correct diff, although it does seem like GH doesn't understand that some lines were just moved)?
IIRC this was a pretty small change, all in all. And I added a few tests to verify correctness + that existing stuff didn't break. That was over two months ago though, so I can revisit if needed.

The question is really whether you're open to PRs for implementing this, or prefer implementing it directly yourself.

> As much as I love this project and want to push it forwards, is not my number one priority 🙏🏼

All good, perhaps it's worth considering adding another mantainer(s) if any are willing to step up. Given that the project's popularity and usage have [skyrocketed](https://npmtrends.com/class-variance-authority) in the past ~year.

**Reply by TomFreudenberg** [reply]
Hi @bengry 

I will try your branch while working on a function that can expose the cva declared variants into the tailwind theme. This will allow easy use of those classes:

### myvariants.ts

```node
const button_variants = cva(["font-semibold", "border", "rounded"], {
  variants: {
    intent: {
      primary: [...]
      secondary: [...]
    },
    size: {
      small: [...],
      medium: [...],
      large: [...],
    },
  },
  compoundVariants: [
    {
      intent: "primary",
      size: "medium",
      className: "uppercase"
    },
  ],
  defaultVariants: {
    intent: "primary",
    size: "medium",
  },
});
```

### tailwind.config.ts

```node
addUtilities(
  cvaExpose('myprefix', {
    'button': button_variants,
    'card': dialog_variants,
    ...
  })
);
```

Can then be used inside the react pages.

### pages.tsx

```node

  <div className="myprefix-card-highlight-bold">
    <h1>Fancy</h1>
    <p>Simple design system</p>
    <button className="myprefix-button-primary-large">Press Me!</div>
  </div>

```

and inside .css as well

### newstyles.css

```css
.button-red {
  @apply myprefix-button-primary-large;
  @apply bg-red-500;
}
```

Feels like complete cva extended tailwind themes?!

Cheers
Tom

P.S.: cvaExpose will also use twMerge to allow merging of cva's for styles as well.

**Reply by joe-bell** [reply]
> The question is really whether you're open to PRs for implementing this, or prefer implementing it directly yourself.

The fact is more down to the fact that implementing features such as this require careful thought and planning — regardless of who creates them. Skyrocketed usage brings skyrocketed responsibility.

Bringing on more contributors also carries responsibility 

I am interested and I will update you when I have the opportunity to evaluate. All I ask for is a little patience!

That’s all I have to say on this for now 🙏🏼

**Reply by elbasan** [reply]
any updates on this?

**Reply by bengry** [reply]
I've updated [my minor fork](https://github.com/bengry/cva/tree/variants-return-property) to support a more advanced use-case since:
```ts
const box = cva({
  variants: {
    shadow: {
      sm: "shadow-sm",
      md: "shadow-md"
    }
  },
  defaultVariants: {
    shadow: "sm"
  }
});

box.variants;
//      ^?
//      { shadow: ReadonlyArray<{ 'sm' | 'md' }> }
```

This allows using [`objectKeys(box.variants)`](https://github.com/bengry/typedash/blob/7a0f4b129b3eb2cf862c8e8a0366ddbb7a1266a3/src/functions/objectKeys/objectKeys.ts#L7-L17) to get the names of all variants, while `box.variants.shadow` is also useful in itself, for things like [Storybook's `argTypes` and `args`](https://storybook.js.org/docs/writing-stories/args#mapping-to-complex-arg-values) (when using controls that require `options`).

We've been using this in production for a few months now and find it very useful.

P.S.
All of the above also works (and is typed accordingly) for `compose`.
Relevant [tests](https://github.com/bengry/cva/blob/main/packages/cva/src/index.test.ts) were also added to the fork.

Do note that all of this is a fork for `cva@beta`, not `class-variance-authority`.

**Reply by joe-bell** [reply]
Hey @bengry, finally finding some time to come back to this. Really impressed by your `.variants` feature! Would you be OK for me to take some of this work as inspiration for the official solution? More than happy to provide credit/co-author commits where necessary

**Reply by bengry** [reply]
@joe-bell What timing. I _just_ published a [`cva-extended` package](https://www.npmjs.com/package/cva-extended) as a [proper fork](https://github.com/bengry/cva-extended).

It does a bit more than just the `.variants`. You can see the changes made to the original package [listed here](https://github.com/bengry/cva-extended/tree/main/packages/cva-extended#extended-features).

We've been using all of these at my day job for the past year or more and it has been working great for us.

Feel free to take any inspiration from it and merge things back to `cva`. I think we'll still need the fork though, unless all of these features land in `cva` itself. Credit and co-author commits would be appreciated :)

**Reply by joe-bell** [reply]
@bengry ha amazing timing, apologies for the massive delay on my end. Congrats on the ship!

Thanks for the overview of changes, I’ll of course make sure you’re credited where credit is due. I’ll keep you posted on where I end up 🤝 

Your extended feature list is eerily close to the private brain-dump I’ve been jotting away in for months; I think the only difference is I would’ve opted for `_` instead `$` for private variants. 

I’m still a little torn on required/optional variants as it’s proved fiddly when needing to reference default props in framework components such as React, but maybe this could be solved by exporting a `.defaultVariants` for those use-cases 🤔 

Happy to see that we’re pretty aligned, and excited to hear how your fork continues

**Reply by bengry** [reply]
@joe-bell all good. Since I forked the package a while ago (~`1.0.0.beta.1`, over a year ago at this point) we've been working with it, so no worries. It also allowed us to experiment with other changes - like making the props required when no `defaultVariants` is specified for a given variant, as well as private variants. Both have proved themselves, especially the former.

> Thanks for the overview of changes, I’ll of course make sure you’re credited where credit is due. >  I’ll keep you posted on where I end up 🤝

Thanks, much appreciated :), and please do - I want to see what ends up being the change and whether or not we need to keep the fork around, or at least how to rebase on top of your work :)

> Your extended feature list is eerily close to the private brain-dump I’ve been jotting away in for months; I think the only difference is I would’ve opted for `_` instead `$` for private variants.

I went with `$` since it made things _a tiny bit_ easier to read as object keys IMO, but really - I don't care either way. As long as the ability is there and is easy to use.

> I’m still a little torn on required/optional variants as it’s proved fiddly when needing to reference default props in framework components such as React, but maybe this could be solved by exporting a `.defaultVariants` for those use-cases 🤔

We're using React (with the component's props usually just `extend`-ing `VariantProps` (or `&`-ing), as usual. It has worked fine for us so far in dozens of components, and since we've added the stricter types we found things are easier to reason about and understand what's going on - things are more explicit.

In what cases do you need to reference the default prop value? These two options for APIs exposing this come to mind, in any case:
```ts
const button = cva({
  variants: {
    intent: {
      primary: "...",
      secondary: "...",
    },
  },
  defaultVariants: {
    intent: "primary",
  },
});

// Option #1
button.variants.intent // ["primary", "secondary"]
button.variants.intent.default // "primary"
expectTypeOf<button.variants.intent>.toEqual<ReadonlyArray<'primary' | 'secondary'> & {
  default: 'primary',
}>();

// Option #2
button.defaultVariants.intent // "primary"
```

> Happy to see that we’re pretty aligned, and excited to hear how your fork continues

No further plans at the moment beyond the three things I added. I just published it since it looks like it could be useful for others. The timing couldn't be more coincidental though 😄

**Comment by TaraShevchenko** [comment]
Also came across a need for similar functionality (getting a config or variants). I am looking forward to the update with the addition of a feature adding this functionality. Thanks more for a great tool!

**Comment by fmonper1** [comment]
This would be a killer feature since it would ensure all our stories are automatically in sync with the variants we define

**Comment by bengry** [comment]
I've published a fork of this package that adds support for this, as well as a couple of other things we've needed using `cva` under the [`cva-extended`](https://www.npmjs.com/package/cva-extended) package.
Do note that it's a breaking change from `cva@1.x` since it has [stricter types](https://github.com/bengry/cva-extended/tree/main/packages/cva-extended#require-variants-unless-a-default-is-provided-in-defaultvariants). Otherwise, it's a superset of `cva`'s capabilities. It's published as `cva-extended@2` just because of that, to avoid confusion with `cva@1`.

**Comment by joe-bell** [comment, marked as answer]
Tackling in #333!
